src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs

```diff
@@ -479,357 +487,304 @@ public Task<Compilation> GetCompilationAsync(SolutionCompilationState compilatio
-            {
-                try
+                // <summary>
+                // Add all appropriate references to the compilation and set it as our final compilation
+                // state.
+                // </summary>
+                // <param name="compilationWithStaleGeneratedTrees">The compilation from a prior run that contains
+                // generated trees, which match the states included in <paramref name="inProgressState"/>'s
+                // generator info. If a generator run here produces the same set of generated documents as are in <paramref
+                // name="inProgressState"/>'s generator info, and we don't need to make any other changes to
+                // references, we can then use this compilation instead of re-adding source generated files again to the
+                // compilation that <paramref name="inProgressState"/> points to.</param>
+                async Task<FinalState> FinalizeCompilationAsync(
+                    InProgressState inProgressState)
                 {
-                    // The final state we produce will be frozen or not depending on if a frozen state was passed into it.
-                    var isFrozen = allSyntaxTreesParsedState.IsFrozen;
-                    var generatorInfo = allSyntaxTreesParsedState.GeneratorInfo;
-                    var compilationWithoutGeneratedDocuments = allSyntaxTreesParsedState.CompilationWithoutGeneratedDocuments;
-
-                    // Project is complete only if the following are all true:
-                    //  1. HasAllInformation flag is set for the project
-                    //  2. Either the project has non-zero metadata references OR this is the corlib project.
-                    //     For the latter, we use a heuristic if the underlying compilation defines "System.Object" type.
-                    var hasSuccessfullyLoaded = this.ProjectState.HasAllInformation &&
-                        (this.ProjectState.MetadataReferences.Count > 0 ||
-                         compilationWithoutGeneratedDocuments.GetTypeByMetadataName("System.Object") != null);
-
-                    var newReferences = new List<MetadataReference>();
-                    var metadataReferenceToProjectId = new Dictionary<MetadataReference, ProjectId>();
-                    newReferences.AddRange(this.ProjectState.MetadataReferences);
-
-                    foreach (var projectReference in this.ProjectState.ProjectReferences)
+                    try
                     {
-                        var referencedProject = compilationState.SolutionState.GetProjectState(projectReference.ProjectId);
-
-                        // Even though we're creating a final compilation (vs. an in progress compilation),
-                        // it's possible that the target project has been removed.
-                        if (referencedProject != null)
+                        Contract.ThrowIfFalse(inProgressState.IntermediateProjects.IsEmpty);
```

Comment:
this assertion is now only true once we've collapsed the whole inprogressstate translation chain.